From ed8480d26dc1d720bdca07d348686296cdad4d0d Mon Sep 17 00:00:00 2001
From: "Lu.Jiang" <lu.jiang@windriver.com>
Date: Fri, 5 Sep 2014 16:24:24 +0800
Subject: [PATCH 1147/1207] powerpc/e6500: Rename PPC_E6500_REV1_BUGS to
 FSL_ERRATUM_A_05337

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

Also make it default y so that people who don't update their config
won't get caught by the late config change.

Signed-off-by: Scott Wood <scottwood@freescale.com>
Change-Id: I28d5752dec9ce1c65e8d50dcbd9c2771d57184e2
Reviewed-on: http://git.am.freescale.net:8181/2559
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Tested-by: Fleming Andrew-AFLEMING <AFLEMING@freescale.com>
Signed-off-by: Lu.Jiang <lu.jiang@windriver.com>
---
 arch/powerpc/mm/tlb_nohash.c           |  7 +++++++
 arch/powerpc/platforms/Kconfig.cputype | 11 +++++++++++
 2 files changed, 18 insertions(+)

diff --git a/arch/powerpc/mm/tlb_nohash.c b/arch/powerpc/mm/tlb_nohash.c
index cbd3d06..890e7b2 100644
--- a/arch/powerpc/mm/tlb_nohash.c
+++ b/arch/powerpc/mm/tlb_nohash.c
@@ -57,8 +57,13 @@
 struct mmu_psize_def mmu_psize_defs[MMU_PAGE_COUNT] = {
 	[MMU_PAGE_4K] = {
 		.shift	= 12,
+		.ind	= 21,
 		.enc	= BOOK3E_PAGESZ_4K,
 	},
+	[MMU_PAGE_1M] = {
+		.shift	= 20,
+		.enc	= BOOK3E_PAGESZ_1M,
+	},
 	[MMU_PAGE_2M] = {
 		.shift	= 21,
 		.enc	= BOOK3E_PAGESZ_2M,
@@ -472,6 +477,7 @@ static void setup_page_sizes(void)
 		tlb1ps = mfspr(SPRN_TLB1PS);
 		eptcfg = mfspr(SPRN_EPTCFG);
 
+#ifndef CONFIG_FSL_ERRATUM_A_005337
 		if ((tlb1cfg & TLBnCFG_IND) && (tlb0cfg & TLBnCFG_PT))
 			book3e_htw_mode = PPC_HTW_E6500;
 
@@ -482,6 +488,7 @@ static void setup_page_sizes(void)
 		 */
 		if (eptcfg != 2)
 			book3e_htw_mode = PPC_HTW_NONE;
+#endif
 
 		for (psize = 0; psize < MMU_PAGE_COUNT; ++psize) {
 			struct mmu_psize_def *def = &mmu_psize_defs[psize];
diff --git a/arch/powerpc/platforms/Kconfig.cputype b/arch/powerpc/platforms/Kconfig.cputype
index 4f1436e..f8c1af9 100644
--- a/arch/powerpc/platforms/Kconfig.cputype
+++ b/arch/powerpc/platforms/Kconfig.cputype
@@ -177,6 +177,17 @@ config PPC_E500MC
 	  such as e5500/e6500), and must be disabled for running on
 	  e500v1 or e500v2.
 
+config FSL_ERRATUM_A_005337
+	bool "Work around erratum A-005337 (no hw tablewalk)"
+	depends on PPC_E500MC
+	default y
+	help
+	  This works around erratum A-005337 by not using hardware tablewalk,
+	  even if the hardware advertises it as present.
+
+	  Say Y if if you need to be able to run on rev1 silicon, otherwise
+	  say N for better performance.
+
 config FSL_ERRATUM_A_006184
 	bool "Work around erratum A-006184"
 	help
-- 
2.0.2

